src/wazuh_modules/syscollector/syscollector_bsd.c

```diff
@@ -1349,6 +1353,224 @@ int getGatewayList(OSHash *gateway_list){
+            break;
+        case TSI_S_LISTEN:
+            snprintf(port_state, STATE_LENGTH, "%s", "listening");
+            break;
+        case TSI_S_CLOSING:
+            snprintf(port_state, STATE_LENGTH, "%s", "closing");
+            break;
+        default:
+            snprintf(port_state, STATE_LENGTH, "%s", "unknown");
+            break;
+    }
+    return port_state;
+}
+
+void sys_ports_mac(int queue_fd, const char* WM_SYS_LOCATION, int check_all){
+    int random_id = os_random();
+    char *timestamp;
+    struct tm localtm;
+
+    time_t now = time(NULL);
+    localtime_r(&now, &localtm);
+
+    os_calloc(TIME_LENGTH, sizeof(char), timestamp);
+
+    snprintf(timestamp, TIME_LENGTH - 1, "%d/%02d/%02d %02d:%02d:%02d",
+            localtm.tm_year + 1900, localtm.tm_mon + 1,
+            localtm.tm_mday, localtm.tm_hour, localtm.tm_min, localtm.tm_sec);
+
+    if (random_id < 0) {
+        random_id = -random_id;
+    }
+
+    // Define time to sleep between messages sent
+    int usec = 1000000 / wm_max_eps;
+
+    mtdebug1(WM_SYS_LOGTAG, "Starting ports inventory.");
+
+    pid_t * pids = NULL;
+    int32_t maxproc;
+    size_t len = sizeof(maxproc);
+    sysctlbyname("kern.maxproc", &maxproc, &len, NULL, 0);
+
+    os_calloc(maxproc, 1, pids);
+    int count = proc_listallpids(pids, maxproc);
+
+    int index;
+
+    for(index=0; index < count; ++index) {
+        pid_t pid = pids[index];
+        // Figure out the size of the buffer needed to hold the list of open FDs
+        int bufferSize = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, 0, 0);
+        if (bufferSize == -1) {
+            mterror(WM_SYS_LOGTAG, "Unable to get open file handles for %d", pid);
+            continue;
+        }
+
+        // Get the list of open FDs
+        struct proc_fdinfo *procFDInfo = (struct proc_fdinfo *)os_malloc(bufferSize);
+        if (!procFDInfo) {
+            mterror(WM_SYS_LOGTAG, "Out of memory. Unable to allocate buffer with %d bytes", bufferSize);
+            continue;
+        }
+        proc_pidinfo(pid, PROC_PIDLISTFDS, 0, procFDInfo, bufferSize);
+        int numberOfProcFDs = bufferSize / PROC_PIDLISTFD_SIZE;
+
+        int i;
+        for(i = 0; i < numberOfProcFDs; i++) {
+            if(procFDInfo[i].proc_fdtype == PROX_FDTYPE_SOCKET) {
+                struct  proc_bsdinfo pbsd;
+                proc_pidinfo(pid, PROC_PIDTBSDINFO, 0, &pbsd, PROC_PIDTBSDINFO_SIZE);
+                struct socket_fdinfo socketInfo;
+                int bytesUsed = proc_pidfdinfo(pid, procFDInfo[i].proc_fd, PROC_PIDFDSOCKETINFO, &socketInfo, PROC_PIDFDSOCKETINFO_SIZE);
+
+                if (bytesUsed == PROC_PIDFDSOCKETINFO_SIZE) {
+                    int localPort = 0;
+                    int remotePort = 0;
+                    char *protocol;
+                    char laddr[NI_MAXHOST];
+                    char faddr[NI_MAXHOST];
+                    int listening = 0;
+
+                    if(socketInfo.psi.soi_family == AF_INET) {
```

Comment:
Review `struct sockaddr_storage` an make this as most AF-independent as possible.